sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1811,7 +1811,7 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
     case Filter(condition, project @ Project(fields, grandChild))
       if fields.forall(_.deterministic) && canPushThroughCondition(grandChild, condition) =>
       val aliasMap = getAliasMap(project)
-      project.copy(child = Filter(replaceAlias(condition, aliasMap), grandChild))
+      project.copy(child = Filter(rewriteCondition(condition, aliasMap), grandChild))
```

Comment:
I think we should rewrite the `Project` here. A predicate with `With` expression means an associated `Project` to output the common expressions. We should rewrite the project expressions that are referenced by the predicate into `Attribute`.

Reply:
If the filter is pushed down multiple nodes, the attribute will be lost.

Reply:
This also means that when we push down predicate through other operators, they should propagate the attributes. Looking at `def canPushThrough`, I think most of them are fine (they propagate all output attribute from child), but `Expand` need update, we need to append the attributes that will be generated by `With` to `Expand#projections` and `output`.

Reply:
Aggregation is also need update, but groupingExpressions will be changed, which may cause regression.

Reply:
can you elaborate about the possible regression?

Reply:
Introduce complex types or variable length types group keys, before is HashAgg, after is SortAgg